sys/init.c

```diff
@@ -368,31 +423,313 @@ InsertMountEntry(PDOKAN_GLOBAL dokanGlobal, PDOKAN_CONTROL DokanControl) {
 
   InitializeListHead(&mountEntry->ListEntry);
 
-  ExAcquireResourceExclusiveLite(&dokanGlobal->Resource, TRUE);
+  if (lockGlobal) {
+      ExAcquireResourceExclusiveLite(&dokanGlobal->Resource, TRUE);
+  }
+
   InsertTailList(&dokanGlobal->MountPointList, &mountEntry->ListEntry);
-  ExReleaseResourceLite(&dokanGlobal->Resource);
 
+  if (lockGlobal) {
+      ExReleaseResourceLite(&dokanGlobal->Resource);
+  }
   return mountEntry;
 }
 
+VOID DeleteDeviceDelayed(PDOKAN_GLOBAL dokanGlobal) {
+    
+    
+    PDEVICE_ENTRY deviceEntry;
+    LIST_ENTRY completeList;
+    PLIST_ENTRY listHead = &dokanGlobal->DeviceDeleteList;
+    PLIST_ENTRY entry;
+    PLIST_ENTRY nextEntry;
+    ULONG totalCount = 0;
+    PDokanDCB dcb;
+
+    InitializeListHead(&completeList);
+
+    if (!ExAcquireResourceExclusiveLite(&dokanGlobal->Resource, FALSE)) {
+        DDbgPrint("  Not able to aquire dokanGlobal->Resource \n");
+        return;
+    }
+    
+    if (IsListEmpty(&dokanGlobal->DeviceDeleteList)) {
+        ExReleaseResourceLite(&dokanGlobal->Resource);
+        return;
+    }
+
+    for (entry = listHead->Flink, nextEntry = entry->Flink;
+        entry != listHead;
+        entry = nextEntry, nextEntry = entry->Flink, totalCount++) {
+
+        deviceEntry = CONTAINING_RECORD(entry, DEVICE_ENTRY, ListEntry);
+        if (deviceEntry) {
+            DDbgPrint("  There is a device for delayed delete. Counter %lu \n", deviceEntry->Counter);
+            InterlockedIncrement((LONG *)&deviceEntry->Counter);
+            
+            BOOLEAN canDeleteDiskDevice = deviceEntry->DiskDeviceObject->ReferenceCount == 0;
+
+            dcb = deviceEntry->DiskDeviceObject->DeviceExtension;
+
+            if (deviceEntry->Counter > 3) {
```

Comment:
Will add some notes. It would no be really delayed if we wouldn't have such value. But you are right